---
title: BMEG 424 Assignment 4
output:
  github_document:
    toc: true
    toc_depth: 4
---

```{r setup, incldue=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# BMEG 424 Assignment 4: ATAC-seq

## Introduction:

### General Information About Assignments

We would like to begin by reminding you of some of the key rules and regulations you will have to follow for the assignments in this course:

-   All assignments are to be completed individually or with one partner

-   If you choose to work in pairs for an assignment, **ONE** of you should make a submission on canvas.
    In the submission you **MUST** report the name & student number of BOTH partners.

-   Each time you work with a new partner you will need to create a new GitHub repo for that assignment.
    You should share your repos with the @BMEGGenoInfo account on Github so TAs have access to your submission.

-   Late assignments will have 10% deducted for each day they are late and assignments will not be accepted \>3 days late.
    Making changes to the files on your repo after the assignment has been submitted on Canvas will result in a late penalty unless an instructor has specifically asked you to change something.

-   Each student will have to ensure their assignment files are knit correctly (see below for instructions) and legible on GitHub.
    20% or more of the assignment marks can be removed at the TAs discretion for incorrect knitting or formatting of assignment files.

-   We are aware that some of you will choose to use large language models like ChatGPT to complete some portions of the assignment.
    You are free to use ChatGPT or other LLMs to help you on the assignment, *provided you cite the model and a link to the conversation* (see [here](https://help.openai.com/en/articles/7925741-chatgpt-shared-links-faq) for how to do this with ChatGPT)

-   More generally, it is a requirement that you include any and all sources from where you received help (excluding your previously mentioned partner) or information for completing any question on the assignments.
    If you are unsure of when/how to cite your sources (including LLMs), please get in contact with the instructors.

If any of these rules are unclear please get in contact with the instructors.

### Goals and Objectives

The goal of todays assignment is to familiarize you with ATAC-seq data.
You recently learned about ATAC-seq in class and how it can be used to understand DNA accessibility across different genomic regions and under different conditions.
In the last assignment we mainly focused on processing our ChIP-seq data (ex. calling peaks) and some simple analysis on processed data which could be used to understand the nature of the data and experiments they came from.

In this assignment we will be working with ATAC-seq data files which have been almost completely processed and will instead focus much more on the analysis of the data (instead of processing).

### Data

The data for todays assignments are located at `/projects/bmeg/A4/`.
None of your processing needs to be done on the server today so you should be able to download the file (yes, singular!) directly to your local machine.

### Software and tools

As stated above, your data is already processed so you will not need to use any of the command line tools used in the previous assignments.
Instead, you will be using R and RStudio to perform your analysis.
You will need to install a few packages to your R environment to complete this assignment.
You may have some of these installed from previous assignments, if not install them now.
The packages are:

-   [tidyverse](https://www.tidyverse.org/) (includes tidyr, dplyr, ggplot2, etc.)

-   [edgeR](https://www.bioconductor.org/packages/devel/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)

-   [GenomicRanges](https://bioconductor.org/packages/devel/bioc/vignettes/GenomicRanges/inst/doc/GenomicRangesIntroduction.html)

-   [Biostrings](https://bioconductor.org/packages/devel/bioc/manuals/Biostrings/man/Biostrings.pdf)

-   [csaw](https://bioconductor.org/packages/release/bioc/html/csaw.html)

-   [gridExtra](https://cran.r-project.org/web/packages/gridExtra/index.html) (only needed for the MA plots)

### Submission

Submit your assignment as a **knitted** RMarkdown document.
*Remember to specify the output as github_document* You will push your knitted RMarkdown document to your github repository (one for each group).
Double check that all files (including figures) necessary for your document to render properly are uploaded to your repository.

You will then submit the link, along with the names and student numbers of all students who worked on the assignment to the assignment 3 page on Canvas. Your assignment should be submitted, and your last commit should be made, before 11:59pm on the day of the deadline. Late assignments will will be deducted 10% per day late. Assignments will not be accepted after 3 days past the deadline.
**NOTE**: This assignment contains A LOT of marks attributed to figures looking right. Please make sure those figures are viewable on GitHub before the deadline.

## Experiment and Analysis:

### 1. ATAC-seq

Our data for this assignment comes from a study which examined the effect of pharmacological inhibition of the BAF chromatin-remodelling complexes.
Mutations in these complexes are often associated with drastic changes in chromatin accessibility and, as a result, cancer formation.
In the study, the authors used drugs to mimic the effects of these cancer-associated mutations and measured the effect on chromatin accessibility using ATAC-seq.

### 2. Exploratory Data Analysis and QC

#### a. Load and reformat data

Download the data file from `/projects/bmeg/A4/` and load it into R using `read.table`, the code for this is provided below:

```{r}
# Read in gzipped file from working directory
downloaded_file_path <- "remote/400E_Assignment_Keys/W2025/A4/count_matrix_raw_atac_BRM014_ACBI1.csv.gz"
atacSeqData <- read.table(gzfile(downloaded_file_path), sep = ",", stringsAsFactors = FALSE, header = TRUE)
```

You can use `head` and `tail` to look at the first and last few rows of the data.

```{r}
head(atacSeqData)
tail(atacSeqData)
```

You can use the `dim` function to see the dimensions of your data, and the `names` function to see the column names.

```{r}
dim(atacSeqData)
names(atacSeqData)
```

Assuming you have not somehow messed up the download, you should have a data frame with 25 columns and 56617 rows.
The first column is the genomic region while the other columns are different experiments that were carried out.
Your columns should be formatted like this:

```         
R1_24h_DMSO
```

Here, R1 represents the replicate number (each experiment was run in duplicate), 24h represents the time point (the effect of the different drugs tested was measured at multiple time points) and DMSO represents the drug used.
There are two different drugs used in this study, BRM014 (labelled "N" in our data file) and BI_protac, DMSO is one of the control conditions (there is another control condition with no drug added labelled "control").
This means that there are 4 different conditions in this study: BRM014, BI_protac, DMSO and control.
**You'll want to rename all of your columns which currently include "N" to include "BRM014" instead, as it will make the rest of this assignment a lot less confusing.** Another good idea is to **rename "BI_protac" to "BI-protac"** to avoid confusion with the "\_" delimiter that separates the replicate, timepoint and treatment.

```{r}
# rename rows with "N" to "BRM014"
colnames(atacSeqData) <- gsub("N", "BRM014", colnames(atacSeqData))
# rename rows with "BI_protac" to "BI-protac" to avoid confusion with the "_" delimiter
colnames(atacSeqData) <- gsub("BI_protac", "BI-protac", colnames(atacSeqData))
```

As you can see, the data are currently stored in "wide" format, where each column represents a different experiment.
While this format is pretty intuitive for humans to read, it is not ideal for analysis.
We will want to convert our data to "tidy" format where each row represents an individual observation (currently a single row in our data represents 24 different data points/observations).
Our data will become much "longer" as we will have 24 rows for each row in our current data frame.
After reshaping our current data frame, we will have 56617\*24 = 1,358,808 rows and 5 columns.
The data should look something like this:

```{bash, eval=FALSE}
region,replicate,timepoint,treatment,read_count
chr1:181169-181759,R1,24h,DMSO, 32
chr1:181169-181759,R1,24h,BRM014, 18

...


```

#?# 1. Format your data into tidy format like the example above (0.5 pts)

```{r}

```

#### b. Exploratory Data Analysis

#?# 2. In order to understand your data better create a heatmap of the read counts for each condition. (0.5 pts) 
Have one subplot for each replicate, with timepoints on the y-axis, treatments on the x-axis, and tiles coloured by read count. You should include the number of reads on each tile as well. Note that your heat-map should plot ALL possible permutations of replicate-timepoint-treatment, even if missing from the data file.
(Example: There is no R1_6h_DMSO column in the original data but that tile should exist in your plot)

```{r}

```

#?# 3. Interpret the plot above, what do you notice about the data? (1 pts)
```

```

For the rest of this assignment we are only interested in the BRM014, BI_protac and control conditions.
We will only compare them at the 24h timepoint (this is the only timepoint where all three conditions were measured).
Using the tidy-formatted data you generated in above (in the format like shown on lines 88-92):

#?# 4. Filter your data to only include the BRM014, BI_protac and control conditions at the 24h timepoint. **Keep both replicates**. (0.5 pts)

```{r}

```

#### c. Normalization

We will be using the counts-per-million (CPM) normalization method on our data.
There are far more complex ways to normalize data but CPM is a simple and effective method which serves our purposes well here.
You can run the following code on your filtered data (from Q4) to normalize your data:

```{r}
# atacSeqDataSOI <- replace with <your tidy data frame>
atacSeqDataSOI <- atacSeqDataSOI %>%
    group_by(replicate, timepoint, treatment) %>%
    mutate(countsPerMillion = 1000000 * read_count / sum(read_count))
```

#?# 5. What does CPM normalize for? Is this necessary for our data? How do you know? (1 pts)
```

```

#?# 6. Are there some *non-normalization* methods that would be good alternatives to CPM? Discuss potential advantages and disadvantages of these methods. (1 pts)
```

```

#?#
7.
Using ggplot, plot the kernel-density estimate of the distribution of CPM normalized read counts for each condition.
Use different colours for each curve.
(0.5 pts)

```{r}
# HINT: You should log-scale your data when plotting the KDE

```

#?# 8. Interpret the plot above, what do you notice about the data? (0.5 pts)
```

```

#?# 9. Filter any regions which have a CPM (in any condition) of >400 into a separate data.frame and inspect them. What do you notice about these regions? (1.5 pts)

```{r}
# HINT: Once you've generated your new data frame you can view it as a table in RStudio by clicking on the data frame in the environment tab and then clicking "view" or export it to a csv file using the write.table function

```

#### d. Filtering

We will now filter our data to remove regions which have very low read counts.
This is a common step in ATAC-seq analysis as regions with very low read counts are often not real and are instead the result of noise in the data.
We will remove regions which have a mean of less than 10 reads.

#?# 10. Filter your data to remove regions which have a mean of less than 10 reads across our samples of interest. (0.5 pts)
```{r}

```

#?# 11. Remove any mitochondrial regions from your data. (0.5 pts)
```{r}

```

### 3. Differential Accessibility Analysis

We will now perform differential accessibility analysis on our data.
We will be using the edgeR package to do this.
edgeR is a package which is used to identify differentially expressed genes in RNA-seq data.
We will be using it to identify differentially accessible regions in our ATAC-seq data.

edgeR will require the data to be back in wide format for the analysis. We will need to convert our data back into wide format before we can continue. 
Your data should look like this (there are no values in the table below, this is just to show you the general structure):
```
                        R1_24h_control	R1_24h_BI-protac	R2_24h_control	R2_24h_BI-protac 
chr4:84336780-84337568   
chr8:112059072-112060092 
chr17:55482241-55483505

```

Below we've provided a function which will:

1. Format your data like above
   
2. Perform differential accessibility analysis on your data. You can use this function to perform differential accessibility analysis on your data.

**NOTE**: This function may not work as is, if you've done any of the above steps incorrectly or even if you've done them 'correctly' but in a way we didn't expect, you may need to modify the function to get it to run with your own data. A common error in previous years was the datatype being incorrect.
```{r}
library(csaw)
library(edgeR)

analyze_differential_expression <- function(atacSeqData, comparison, normalization_method = "TMM") {
    # Format data (cast to wide format) and isolate the columns we need based on the "comparison" argument
	# This block of code is highly likely to fail if your data is formatted differently to how we expected  
    countsMatrix <- atacSeqData %>%
        filter(treatment %in% comparison) %>%
        mutate(condition = paste(replicate, timepoint, treatment, sep = "_")) %>%
        select(region, condition, countsPerMillion) %>%
        pivot_wider(names_from = condition, values_from = countsPerMillion) %>%
        column_to_rownames(var = "region")

    treatments <- gsub("R\\d+_\\d+h_(\\w+)", "\\1", colnames(countsMatrix))
    treatments <- factor(treatments)

    y <- DGEList(counts = countsMatrix, group = treatments)

    if (normalization_method == "loess") {
        y <- normOffsets(y, method = "loess")
    } else if (normalization_method == "TMM") {
        y <- calcNormFactors(y)
    }

    design <- model.matrix(~treatments)
    print(design)

    y <- estimateDisp(y, design)

    # Perform differential expression analysis
    fit <- glmQLFit(y, design)
    qlf <- glmQLFTest(fit, coef = 2)

    # Get the top differentially expressed genes
    allDEStats <- as.data.frame(topTags(qlf, n = nrow(countsMatrix)))
    allDEStats$region <- row.names(allDEStats)

    return(allDEStats)
}

# Example usage:
# result <- analyze_differential_expression(atacSeqDataSOI, comparison=c("BRM014","BI_protac"), normalization_method="loess")
```

You will need to provide the function with your data (the tidy and filtered atac-seq data) and the comparison you want to perform.
You should perform each of the following comparisons:

-   Control vs BI_protac
-   Control vs BRM014
-   BRM014 vs BI_protac

#?# 12. Run the above function *six* times to perform each of the comparisons listed above. You should perform each comparison with loess and TMM normalization (1 pts)
```{r}

```

#?# 13. Create 6 MA plots (using ggplot) for each of the comparisons you performed in the previous question. Present them as subplots of the same figure. (1 pts)
```{r}
# HINT: There are many packages which implement MA plots in R, you are supposed to use ggplot to make them manually.
library(gridExtra) # This package will be useful for creating subplots

```

#?# 14. Create (using ggplot) a volcano plot for the comparison between BRM014 and BI_protac (loess normalized), with -log10(p-value) on the y axis and logFC on the x. Colour points that are significant at a FDR<0.01. (0.5 pts)
```{r}

```

### 4. "Gene" Ontology (GO) Analysis

For this section we will be using GREAT (Genomic Regions Enrichment of Annotations Tool) which is a tool used to identify the biological functions of a set of genomic regions.
GREAT uses the Gene Ontology (GO) database to identify the biological functions of a set of genomic regions.
The GO database is a collection of terms which describe the biological functions of genes.
For example, the term "cell cycle" describes the function of genes which are involved in the cell cycle.

GREAT works by taking a set of genomic regions and identifying the genes which are nearby to those regions.
It then uses the GO database to identify the biological functions of those genes.
It then uses a hypergeometric test to identify which GO terms are enriched in the set of genes near the genomic regions of interest.
*GREAT is available as a web tool (<https://great.stanford.edu/great/public/html/splash.php>)*

#### a. Preparing your data for GREAT

GREAT requires your data to be in a specific format.
You will need to convert your data to this format before you can use GREAT.
The format is as follows:

```         
chr1    10000   20000
chr1    30000   40000
...
```

As you can see this is the basics of a BED file.
You will need to convert your data to this format.
You can do this using the `write.table` function.
You'll need to separate the "region" column in your differential analysis results into three columns: chromosome, start and end.
You can do this using the `separate` function from the `tidyr` package or using regular expressions (regex).

#?# 15. Create a pseudo-BED file for each of the results of your differential analysis. Fill in the code you used below. (1pts)
```{r, eval=FALSE}

```


#### b. Interpreting your results

Once you have created your pseudo-BED files (one for each of the results of your differential analysis) you can upload them to GREAT.
Use the hg38 genome, with the whole genome being your background.
You will need to use the "single nearest gene" option (click on advanced options) when testing your data.

Because the drugs we are using affect accessibility broadly across the genome our results will be tough for GREAT's hypergeometric test to interpret.
You should select the "Significant By Region-based Binomial view." GREAT should prompt you to do this after it displays your results.
This will be more informative for our data.

Once you have tested your data on GREAT you will be presented with a list of GO terms which are enriched in your data.
You can click on each GO term to see the genes which are associated with that term.
You can also click on the "Genes" tab to see all of the genes which are associated with your data.

#?# 16. Upload your (loess normalized) ControlvsBRM014 and ControlvsBI_protac results to GREAT. Insert a screenshot of the top 20 associated ENSEMBL genes for each comparison. (1 pts)
![]()

![]()



## Discussion
#?# 17. Based on your analysis above, what effect do you think BRM014 and BI_protac had on chromatin accessibility? Are they similar? Use evidence from your analysis to support your response (2 pts) 
```

```

#?# 18. Based on your analysis above, can you tell if BRM014 and/or BI_protac are good pharmacological models for studying the effects of the cancer-associated mutations the authors were trying to mimic? Why or why not? (2 pts)
```
#HINT: Print the length of the bed files you made above, do you notice anything interesting? 

```

#?# 19. We used two normalization methods in our differential accessibility analysis, TMM and Loess. Which do you think is more reliable for our data, and why? (2 pts)
```

```
